### DISCUSSION TITLE
Improvements for displaying errors

### DISCUSSION THREAD

**Discussion by miv391** [godotengine/godot-proposals#4192] [Editor]
Errors are currently displayed in many places. In this example the error is trying to use type inference assignment operator with an array that contains items of multiple types. Error messages are displayed in three different places:

![kuva](https://user-images.githubusercontent.com/49998025/157457192-28910d2b-9181-49ee-bb1b-21f7ce602a0a.png)
![kuva](https://user-images.githubusercontent.com/49998025/157457210-26f035bd-6387-4c3c-ab86-206bc4742c14.png)

(screenshots are from v4.0.alpha4.official [f47097973])

Suggestions: 
- Either display all or no errors in the Output view. I think it would be better to not display any errors in the Output view as there is a dedicated Errors view for that.
- Move the Debugger/Errors view to the top level, next to the Output view. So there would be Output, Errors, Debugger, Search Results etc. That would make the Errors view more easily accessible. Godot is the only IDE I have used where errors are displayed under debugging tools.
- Currently a colored bullet is used to display whether there are warnings and/or errors. If there are both warnings and errors, one half of the bullet is red, other half is yellow. That's cute but useless. It would be better just to show the most severe color. If there are only warnings, use yellow bullet, but if there is even one error, use red bullet.
- The warning/error bullet color works differently in the top level views (Output, Errors, etc.) and in the Debugger sub views (Debugger, Errors, etc.). In the top level, selected and unselected views have different bullet colors. As the screen shot shows, the Debugger view has grey/green bullet (instead of red/yellow) if the Debugger view is selected. The sub level views always have red/yellow colors. It would be better to always use same colors.
![kuva](https://user-images.githubusercontent.com/49998025/157457544-61dc2a5c-1474-4d82-8d4f-07f821cc4e67.png)

**Comment by Calinou** [comment]
> Either display all or no errors in the Output view. I think it would be better to not display any errors in the Output view as there is a dedicated Errors view for that.

I'm not sure if this is a good idea. Not only you lose the ability to see *when* an error was produced relative to standard `print()` messages, but I've seen multiple instances of people missing error messages entirely because they didn't check the **Debugger > Errors** tab.

> Move the Debugger/Errors view to the top level, next to the Output view. So there would be Output, Errors, Debugger, Search Results etc. That would make the Errors view more easily accessible. Godot is the only IDE I have used where errors are displayed under debugging tools.

This makes sense, but the bottom panel can get pretty crowded on narrow monitors already. We'd like the Godot editor to remain usable on laptops with 1366×768 displays.

> Currently a colored bullet is used to display whether there are warnings and/or errors. If there are both warnings and errors, one half of the bullet is red, other half is yellow. That's cute but useless. It would be better just to show the most severe color. If there are only warnings, use yellow bullet, but if there is even one error, use red bullet.

This was done in https://github.com/godotengine/godot/pull/38592. Most people seemed to be in favor of it there :slightly_smiling_face: 

> The warning/error bullet color works differently in the top level views (Output, Errors, etc.) and in the Debugger sub views (Debugger, Errors, etc.). In the top level, selected and unselected views have different bullet colors. As the screen shot shows, the Debugger view has grey/green bullet (instead of red/yellow) if the Debugger view is selected. The sub level views always have red/yellow colors. It would be better to always use same colors.

This is because icons for pressed buttons are modulated. It's likely possible to disable this for specific buttons, but I need to double-check to make sure.

**Reply by miv391** [reply]
> I'm not sure if this is a good idea. Not only you lose the ability to see _when_ an error was produced relative to standard `print()` messages, but I've seen multiple instances of people missing error messages entirely because they didn't check the **Debugger > Errors** tab.

That's a good point (and I might be one of those people who have missed the error messages...) The other option would then be displaying all errors in Output. For example the error caused by this code is currenlty not displayed in Output:

```gdscript
# no error in Output
range(3).map(func (): return)
```

Maybe the goal could be that Ouput contains as clear error messages as possible. For example, let's look at this code 

```gdscript
var arr := [1, 2, 'a']
```

and its current error message:

```
modules/gdscript/gdscript_parser.h:148 - Condition "container_element_type == nullptr" is true. Returning: DataType()  
```

That error message is quite useless. As I already knew what kind of error I was going to produce, I can guess from the error message what's happening in the core, but without that foreknowledge it would be pretty much impossible to know how to fix the error.

Maybe the Debugger/Errors view could be the place for error messages which contain core/C++ specific information?

Suggestion:
- Output has error messages for the game programmer, e.g.: `Cannot use type inferring assignment operator ':=' with a Variant array. Use '=' instead.`
- Debugger/Errors has more detailed error messages mainly targeted for Godot developer team to help fixing bugs in Godot itself: `get_container_element_type: Condition "container_element_type == nullptr" is true. Returning: DataType() <C++ Source>   modules/gdscript/gdscript_parser.h:148 @ get_container_element_type()`


 
> This makes sense, but the bottom panel can get pretty crowded on narrow monitors already. We'd like the Godot editor to remain usable on laptops with 1366×768 displays.

How about changing bottom panel texts to icons when there is not enough space for texts? Although if the Debugger/Errors would contain only error messages mainly targeted the Godot core developers (per my previous suggestion), Errors view would be fine under the Debugger.

 
> This was done in [godotengine/godot#38592](https://github.com/godotengine/godot/pull/38592). Most people seemed to be in favor of it there 🙂

I still think red/yellow bullet is bad UX. Are there ever any warnings that the programmer should really be aware of? Does it matter at all, whether there are both warnings and errors or only errors? Having no bullet, yellow bullet or red bullet are clear and easy to understand. The fourth option, red/yellow bullet, does not give much more useful information that the red bullet alone would. But it adds more cognitive load to the user.
